File: python/x402/schemas/payments.py
"""V2 payment types for the x402 Python SDK."""

from typing import Any

from pydantic import Field

from .base import BaseX402Model, Network


class ResourceInfo(BaseX402Model):
    """Describes the resource being accessed.

    Attributes:
        url: The URL of the resource.
        description: Optional human-readable description.
        mime_type: Optional MIME type of the resource.
        service_name: Optional human-readable service name (≤ 32 chars).
        tags: Optional topical tags for the service (≤ 5 entries, each ≤ 32 chars).
        icon_url: Optional absolute http(s) URL to a service icon (≤ 2048 chars).

    See `specs/extensions/bazaar.md` "Service Metadata on `resource`" for
    facilitator-side validation rules applied to service_name / tags / icon_url.
    """

    url: str
    description: str | None = None
    mime_type: str | None = None
    service_name: str | None = None
    tags: list[str] | None = None
    icon_url: str | None = None


class PaymentRequirements(BaseX402Model):
    """V2 payment requirements structure.

    Attributes:
        scheme: Payment scheme identifier (e.g., "exact").
        network: CAIP-2 network identifier (e.g., "eip155:8453").
        asset: Asset address/identifier.
        amount: Amount in smallest unit.
        pay_to: Recipient address.
        max_timeout_seconds: Maximum time for payment validity.
        extra: Additional scheme-specific data.
    """

    scheme: str
    network: Network
    asset: str
    amount: str
    pay_to: str
    max_timeout_seconds: int
    extra: dict[str, Any] = Field(default_factory=dict)

    def get_amount(self) -> str:
        """Get the payment amount (V2 uses 'amount' field)."""
        return self.amount

    def get_extra(self) -> dict[str, Any] | None:
        """Get extra metadata."""
        return self.extra


class PaymentRequired(BaseX402Model):
    """V2 402 response structure.

    Attributes:
        x402_version: Protocol version (always 2 for V2).
        error: Optional error message.
        resource: Optional resource information.
        accepts: List of accepted payment requirements.
        extensions: Optional extension data.
    """

    x402_version: int = 2
    error: str | None = None
    resource: ResourceInfo | None = None
    accepts: list[PaymentRequirements]
    extensions: dict[str, Any] | None = None


class PaymentPayload(BaseX402Model):
    """V2 payment payload structure.

    Attributes:
        x402_version: Protocol version (always 2 for V2).
        payload: Scheme-specific payload data.
        accepted: The payment requirements being fulfilled.
        resource: Optional resource information.
        extensions: Optional extension data.
    """

    x402_version: int = 2
    payload: dict[str, Any]
    accepted: PaymentRequirements
    resource: ResourceInfo | None = None
    extensions: dict[str, Any] | None = None

    def get_scheme(self) -> str:
        """Get the payment scheme (V2 uses accepted.scheme)."""
        return self.accepted.scheme

    def get_network(self) -> str:
        """Get the network (V2 uses accepted.network)."""
        return self.accepted.network

File: ruby/x402/lib/x402/schemas/payments.rb
# frozen_string_literal: true

require_relative 'base'

module X402
  ##
  # Describes the resource being accessed.
  #
  # @!attribute url
  #   @return [String] the URL of the resource
  # @!attribute description
  #   @return [String, nil] optional human-readable description
  # @!attribute mime_type
  #   @return [String, nil] optional MIME type of the resource
  class ResourceInfo < BaseStruct
    attribute :url, Types::String
    attribute :description, Types::String.optional.default(nil)
    attribute :mime_type, Types::String.optional.default(nil)
  end

  ##
  # V2 payment requirements structure.
  #
  # Describes what payment the server requires to access a resource.
  #
  # @!attribute scheme
  #   @return [String] payment scheme identifier (e.g., "exact")
  # @!attribute network
  #   @return [String] CAIP-2 network identifier (e.g., "eip155:8453")
  # @!attribute asset
  #   @return [String] asset address/identifier
  # @!attribute amount
  #   @return [String] amount in smallest unit
  # @!attribute pay_to
  #   @return [String] recipient address
  # @!attribute max_timeout_seconds
  #   @return [Integer] maximum time for payment validity
  # @!attribute extra
  #   @return [Hash] additional scheme-specific data
  class PaymentRequirements < BaseStruct
    attribute :scheme, Types::String
    attribute :network, Types::String
    attribute :asset, Types::String
    attribute :amount, Types::String
    attribute :pay_to, Types::String
    attribute :max_timeout_seconds, Types::Integer
    attribute :extra, Types::Hash.default({}.freeze)

    ##
    # Get the payment amount (V2 uses 'amount' field).
    #
    # @return [String] the payment amount
    def get_amount
      amount
    end

    ##
    # Get extra metadata.
    #
    # @return [Hash, nil] extra metadata
    def get_extra
      extra.empty? ? nil : extra
    end
  end

  ##
  # V2 402 Payment Required response structure.
  #
  # Sent by server when payment is required to access a resource.
  #
  # @!attribute x402_version
  #   @return [Integer] protocol version (always 2 for V2)
  # @!attribute error
  #   @return [String, nil] optional error message
  # @!attribute resource
  #   @return [ResourceInfo, nil] optional resource information
  # @!attribute accepts
  #   @return [Array<PaymentRequirements>] list of accepted payment requirements
  # @!attribute extensions
  #   @return [Hash, nil] optional extension data
  class PaymentRequired < BaseStruct
    attribute :x402_version, Types::Integer.default(X402_VERSION)
    attribute :error, Types::String.optional.default(nil)
    attribute :resource, ResourceInfo.optional.default(nil)
    attribute :accepts, Types::Array.of(PaymentRequirements)
    attribute :extensions, Types::Hash.optional.default(nil)
  end

  ##
  # V2 payment payload structure.
  #
  # Created by client to prove payment to server.
  #
  # @!attribute x402_version
  #   @return [Integer] protocol version (always 2 for V2)
  # @!attribute payload
  #   @return [Hash] scheme-specific payload data (authorization, signature, etc.)
  # @!attribute accepted
  #   @return [PaymentRequirements] the payment requirements being fulfilled
  # @!attribute resource
  #   @return [ResourceInfo, nil] optional resource information
  # @!attribute extensions
  #   @return [Hash, nil] optional extension data
  class PaymentPayload < BaseStruct
    attribute :x402_version, Types::Integer.default(X402_VERSION)
    attribute :payload, Types::Hash
    attribute :accepted, PaymentRequirements
    attribute :resource, ResourceInfo.optional.default(nil)
    attribute :extensions, Types::Hash.optional.default(nil)

    ##
    # Get the payment scheme (V2 uses accepted.scheme).
    #
    # @return [String] the payment scheme
    def get_scheme
      accepted.scheme
    end

    ##
    # Get the network (V2 uses accepted.network).
    #
    # @return [String] the network identifier
    def get_network
      accepted.network
    end
  end
end
